Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
 
 #include "Chrome.h"
 #include "ChromeClient.h"
+#include "ComposedTreeIterator.h"
 #include "Document.h"
+#include "DocumentType.h"
 #include "Editor.h"
-#include "FindOptions.h"
+#include "HTMLAudioElement.h"
+#include "HTMLIFrameElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLMeterElement.h"
+#include "HTMLObjectElement.h"
+#include "HTMLProgressElement.h"
+#include "HTMLScriptElement.h"
+#include "HTMLStyleElement.h"
+#include "HTMLVideoElement.h"
+#include "Position.h"
 #include "SimpleRange.h"
+#include "TextBoundaries.h"
 #include "TextIterator.h"
 
 namespace WebCore {
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+enum class BoundaryPointIsAtEnd { No, Yes };
+enum class WordBounded : bool { No, Yes };
+
+// https://wicg.github.io/scroll-to-text-fragment/#search-invisible
+static bool isSearchInvisible(const Node& node)
+{
+    // FIXME: The computed value of its display property is none.
+    
+    // FIXME: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node)
+        || is<HTMLImageElement>(node)
+        || is<HTMLMeterElement>(node)
+        || is<HTMLObjectElement>(node)
+        || is<HTMLProgressElement>(node)
+        || is<HTMLStyleElement>(node)
+        || is<HTMLScriptElement>(node)
+        || is<HTMLVideoElement>(node)
+        || is<HTMLAudioElement>(node))
+        return true;
+    
+    // FIXME: Is a select element whose multiple content attribute is absent.
+    
+    return false;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#non-searchable-subtree
+static bool isNonSearchableSubtree(const Node& node)
+{
+    const Node* traversingNode = &node;
+    while (traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentOrShadowHostNode();
+    }
+    return false;
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#nearest-block-ancestor
+static const Node& nearestBlockAncestor(const Node& node)
+{
+    const Node* currentNode = &node;
+    while (currentNode) {
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
```

Comment:
It's unclear whether we should be using flat/composed tree or shadow including tree.
Let's just call parentNode for now to be consistent with elsewhere.